rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+```
+
+#### Error Reporting From `replica_fn`
+
+Because of the non-blocking `schedule`, any exception raised in `replica_fn` wouldn’t be returned to users immediately. Actually an exception may pollute arbitrary number of functions in flight following the culprit function. We will set the error in returned `Future` objects for the culprit function and these polluted functions and we will raise exceptions when `join` is called.
+
+Therefore the best practice for users is to avoid writing any code that may raise in `replica_fn`:
+
+*   use repeated dataset so `OutOfRangeError` will be avoided;
+*   avoid using assertion ops or some debugging ops like `tf.debugging.check_numerics`.
+
+
+#### Dataset Interface
+
+The traditional training loop of `tf.distribute` passes the `get_next` results of a distributed iterator to `replica_fn`:
+
+
+```python
+for x in enumerate(distributed_iter):
+  loss = strategy.schedule(replica_fn, x, y)
+```
+
+
+If we do the same thing with the `strategy.schedule` API, there are several challenges.
+
+The first challenge is we don’t know which worker the `get_next` should return to since where the `replica_fn` will be executed will be decided later. Some later-binding mechanism can be explored.
+
+The second challenge is calling `get_next` on an iterator is synchronous. This means that the training loop is not truly asynchronous. It is tricky to make `get_next` asynchronous because the client doesn’t know how many items will be in the iterator and thus doesn’t know how many functions to schedule. 
+
+
+##### Alternative: passing iterators to `strategy.schedule`
+
+The following training loop is less consistent with other `tf.distribute` examples but is easier to implement in the short term. It requires users to explicitly set a number of steps.
+
+
+```python
+# … omitted
+with strategy.scope():
+  # … omitted
+  distributed_iter = iter(distributed_dataset)
+  for i in range(total_steps):
+    strategy.schedule(replica_fn, args=(distributed_iter,))
+# … omitted
+```
+
+
+**We will start with this kind of training loop in our first version.**
+
+
+### Fault Tolerance
+
+
+#### Task Failure
+
+
+##### Worker failure
+
+
+###### When scheduling
+
+When a worker fails, our training will continue without this failed worker. Functions scheduled on a failed worker will be rescheduled on other workers. 
+
+For functions that bound to a specific worker, e.g. resource creation function, they will be queued until the worker is back.
+
+When the failed worker is back, we will update the cluster configuration with `context.update_server_def` which would also reset all the states. After resources on the restarted worker are built, we can resume scheduling functions on the worker. 
+
+
+###### When materialing a `Future` object
+
+It is possible that a function is executed but its corresponding worker fails when users try to consume its output. In this case, we will give users a `None` value and set an error in the `Future` object.
```

Comment:
Can you show an example of what this would look like, and how a user should write code robust to this?

Reply:
Update the example in the `Custom Training Loop` section. You can see its last line.